--- a/meson.build	2025-11-21 09:25:05.000000000 +0000
+++ b/meson.build	2026-01-08 14:22:53.117984009 +0000
@@ -21,41 +21,3 @@
   'pg_stat_monitor--2.2--2.3.sql',
   kwargs: contrib_data_args,
 )
-
-tests += {
-  'name': 'pg_stat_monitor',
-  'sd': meson.current_source_dir(),
-  'bd': meson.current_build_dir(),
-  'regress': {
-    'sql': [
-      'application_name',
-      'application_name_unique',
-      'basic',
-      'cmd_type',
-      'counters',
-      'database',
-      'different_parent_queries'
-      'error_insert',
-      'error',
-      'functions',
-      'guc',
-      'histogram',
-      'level_tracking'
-      'parallel',
-      'pgsqm_query_id',
-      'relations',
-      'rows',
-      'squashing',
-      'state',
-      'tags',
-      'top_query',
-      'user',
-      'version'
-    ],
-    'regress_args': ['--temp-config', files('pg_stat_monitor.conf')],
-    # Disabled because these tests require
-    # "shared_preload_libraries=pg_stat_monitor", which typical
-    # runningcheck users do not have (e.g. buildfarm clients).
-    'runningcheck': false,
-  },
-}
